app/controllers/lti_controller.rb

```diff
@@ -0,0 +1,81 @@
+class LtiController < ApplicationController
+  before_action :set_group_assignment, only: %i[launch]
+  before_action :set_lti_params, only: %i[launch]
+  after_action :allow_iframe, only: %i[launch]
+  
+  def launch
+    session[:isLTI]=true # the lti session starting
+    require 'oauth/request_proxy/action_controller_request'
+
+    if !@assignment.present?
+      # if no assignment is found
+      flash[:notice] = "No such Assignment is present in CircuitVerse, please contact your LMS Administrator/Teacher"
+      render :launch_error, status: 401
+      return
+    end
+    
+    if @group.present? # if there is a valid group based for the lti_token_key
+      @provider = IMS::LTI::ToolProvider.new(
+        params[:oauth_consumer_key], # lms_oauth_consumer_key
+        @assignment.lti_shared_secret, # the group's lti_token
+        params
+      )
+
+      if not @provider.valid_request?(request) # checking the lti request from the lms end
+        render :launch_error, status: 401
+        return
+      end
+
+      @@launch_params=params;
+      @user = User.find_by(email: @email_from_lms) # find user by matching email with circuitverse and lms 
+
+      if @user.present? # user is present in cv 
+        if @user.id === @group.mentor_id # user is teacher
+          sign_in(@user) # passwordless sign_in the user as the authenticity is verified via lms
```

Comment:
I think you should set LTI session `session[:isLTI]=true`  only after sign in? CC: @tachyons

Reply:
LTI session should not depend on sign in because for student end there is no sign in , and for error cases , say like any teacher added some wrong key and secret then also the session is LTI but they are restricted to access the course